File: src/podcast/routers/pages.py
import json
import uuid

from fastapi import APIRouter, Depends, Request
from fastapi.responses import HTMLResponse, RedirectResponse
from fastapi.templating import Jinja2Templates
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from podcast.auth import require_auth_page
from podcast.database import get_db
from podcast.models import Episode, PodcastSettings
from podcast.services.episode import create_episode, get_episode, list_episodes
from podcast.services.tts import get_tts_progress as _read_tts_progress
from podcast.services.llm_providers import get_all_model_pricing

# Pricing pulled dynamically from provider registry
_MODEL_PRICING = get_all_model_pricing()
# Fallback pricing (Claude Sonnet 4 rates)
_DEFAULT_PRICING = {"input": 3.0, "output": 15.0}

templates = Jinja2Templates(directory="src/podcast/templates")

router = APIRouter()


def _status_badge(status: str) -> str:
    """Map episode status to badge CSS class."""
    mapping = {
        "ready": "badge--success",
        "failed": "badge--error",
        "pending": "badge--info",
    }
    # All in-progress statuses get warning
    return mapping.get(status, "badge--warning")


def _status_label(status: str) -> str:
    """Human-readable status label."""
    mapping = {
        "pending": "Pending",
        "researching": "Researching",
        "writing_transcript": "Writing transcript",
        "generating_audio": "Generating audio",
        "encoding": "Encoding",
        "ready": "Ready",
        "failed": "Failed",
    }
    return mapping.get(status, status.replace("_", " ").title())


def _format_duration(seconds: int | None) -> str:
    if seconds is None:
        return ""
    minutes = seconds // 60
    secs = seconds % 60
    return f"{minutes}:{secs:02d}"


def _format_file_size(size_bytes: int | None) -> str:
    if size_bytes is None:
        return ""
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 * 1024:
        return f"{size_bytes / 1024:.1f} KB"
    else:
        return f"{size_bytes / (1024 * 1024):.1f} MB"


PIPELINE_STEPS = ["research", "transcript", "tts", "encode"]

STEP_LABELS = {
    "research": "Research",
    "transcript": "Transcript",
    "tts": "Audio",
    "encode": "Encode",
}


def _build_pipeline_info(episode) -> list[dict]:
    """Build ordered pipeline step info from episode jobs."""
    jobs_by_step = {job.step: job for job in (episode.jobs or [])}

    steps = []
    for step_name in PIPELINE_STEPS:
        job = jobs_by_step.get(step_name)
        if job:
            duration = None
            if job.started_at and job.completed_at:
                delta = job.completed_at - job.started_at
                total_secs = int(delta.total_seconds())
                if total_secs >= 60:
                    duration = f"{total_secs // 60}m {total_secs % 60}s"
                else:
                    duration = f"{total_secs}s"
            steps.append({
                "name": step_name,
                "label": STEP_LABELS[step_name],
                "status": job.status,
                "attempts": job.attempts,
                "duration": duration,
            })
        else:
            steps.append({
                "name": step_name,
                "label": STEP_LABELS[step_name],
                "status": "waiting",
                "attempts": 0,
                "duration": None,
            })

    return steps


def _get_current_step_index(episode) -> int:
    """Return 0-based index of the current/completed pipeline step."""
    if episode.status == "ready":
        return len(PIPELINE_STEPS)
    jobs_by_step = {job.step: job for job in (episode.jobs or [])}
    for i, step_name in enumerate(PIPELINE_STEPS):
        job = jobs_by_step.get(step_name)
        if not job or job.status in ("pending", "running"):
            return i
    return len(PIPELINE_STEPS)


def _get_tts_progress(episode) -> dict | None:
    """Get TTS progress for an episode, if generating audio."""
    if episode.status != "generating_audio":
        return None
    progress = _read_tts_progress(episode.id)
    if not progress:
        return None
    # Add formatted audio duration for template convenience
    total_secs = int(progress["audio_duration_seconds"])
    minutes = total_secs // 60
    secs = total_secs % 60
    progress["audio_duration_formatted"] = f"{minutes}:{secs:02d}"
    return progress


# Register template globals
templates.env.globals["status_badge"] = _status_badge
templates.env.globals["status_label"] = _status_label
templates.env.globals["format_duration"] = _format_duration
templates.env.globals["format_file_size"] = _format_file_size
templates.env.globals["build_pipeline_info"] = _build_pipeline_info
templates.env.globals["get_current_step_index"] = _get_current_step_index
templates.env.globals["get_tts_progress"] = _get_tts_progress


@router.get("/shoo/callback", response_class=HTMLResponse)
async def shoo_callback(request: Request):
    """Callback page for Shoo OAuth — shoo.js handles the code exchange."""
    return templates.TemplateResponse(request, "auth_callback.html")


@router.get("/", response_class=HTMLResponse)
async def index(request: Request, db: AsyncSession = Depends(get_db)):
    episodes = await list_episodes(db)
    return templates.TemplateResponse(
        request, "index.html", context={"episodes": episodes}
    )


@router.get("/episodes/new", response_class=HTMLResponse)
async def new_episode_page(request: Request, _user: str = Depends(require_auth_page)):
    return templates.TemplateResponse(request, "episode_new.html")


@router.post("/episodes/new")
async def new_episode_submit(request: Request, db: AsyncSession = Depends(get_db), _user: str = Depends(require_auth_page)):
    form = await request.form()
    topic = form.get("topic", "").strip()
    title = form.get("title", "").strip() or None

    # Parse and validate target length
    try:
        target_length = int(form.get("target_length_minutes", "30"))
    except (TypeError, ValueError):
        target_length = 30
    if target_length not in (15, 30, 60, 120):
        target_length = 30

    if not topic:
        return templates.TemplateResponse(
            request,
            "episode_new.html",
            context={"error": "Topic is required"},
        )
    episode = await create_episode(
        db, topic, title,
        target_length_minutes=target_length,
    )
    await db.commit()
    return RedirectResponse(url=f"/episodes/{episode.id}", status_code=303)


@router.get("/episodes/{episode_id}", response_class=HTMLResponse)
async def episode_detail(
    request: Request, episode_id: uuid.UUID, db: AsyncSession = Depends(get_db),
    _user: str = Depends(require_auth_page),
):
    episode = await get_episode(db, episode_id)
    if not episode:
        return HTMLResponse("Not found", status_code=404)
    return templates.TemplateResponse(
        request, "episode_detail.html", context={"episode": episode}
    )


@router.get("/logs", response_class=HTMLResponse)
async def logs_page(request: Request, _user: str = Depends(require_auth_page)):
    return templates.TemplateResponse(request, "logs.html")


@router.get("/settings", response_class=HTMLResponse)
async def settings_page(request: Request, db: AsyncSession = Depends(get_db), _user: str = Depends(require_auth_page)):
    from podcast.services.transcript import DEFAULT_TONE_NOTES

    s = await db.get(PodcastSettings, 1)
    if not s:
        s = PodcastSettings()
        db.add(s)
        await db.flush()

    # Parse stored tone notes (JSON string → list), fall back to defaults
    tone_notes = list(DEFAULT_TONE_NOTES)
    if s.transcript_tone_notes:
        try:
            parsed = json.loads(s.transcript_tone_notes)
            if isinstance(parsed, list):
                tone_notes = parsed
        except (json.JSONDecodeError, TypeError):
            pass

    return templates.TemplateResponse(
        request, "settings.html", context={"settings": s, "tone_notes": tone_notes}
    )


@router.post("/settings")
async def settings_submit(request: Request, db: AsyncSession = Depends(get_db), _user: str = Depends(require_auth_page)):
    form = await request.form()
    s = await db.get(PodcastSettings, 1)
    if not s:
        s = PodcastSettings()
        db.add(s)
        await db.flush()

    for field in ["title", "description", "author", "language", "host_a_name", "host_b_name"]:
        value = form.get(field, "").strip()
        if value:
            setattr(s, field, value)

    return RedirectResponse(url="/settings", status_code=303)


def _calc_cost(input_tokens: int, output_tokens: int, model: str = "") -> float:
    pricing = _MODEL_PRICING.get(model, _DEFAULT_PRICING)
    return (input_tokens * pricing["input"] + output_tokens * pricing["output"]) / 1_000_000


@router.get("/metrics", response_class=HTMLResponse)
async def metrics_page(request: Request, db: AsyncSession = Depends(get_db), _user: str = Depends(require_auth_page)):
    result = await db.execute(
        select(Episode)
        .options(selectinload(Episode.jobs))
        .order_by(Episode.created_at.desc())
    )
    episodes = result.scalars().all()

    episode_rows = []
    totals = {
        "episodes": 0,
        "episodes_ready": 0,
        "total_input_tokens": 0,
        "total_output_tokens": 0,
        "total_cost": 0.0,
        "total_audio_seconds": 0,
        "total_generation_seconds": 0.0,
        "total_tts_seconds": 0.0,
    }

    for ep in episodes:
        totals["episodes"] += 1
        if ep.status == "ready":
            totals["episodes_ready"] += 1

        row = {
            "id": str(ep.id),
            "title": ep.title,
            "status": ep.status,
            "episode_number": ep.episode_number,
            "audio_duration_seconds": ep.audio_duration_seconds,
            "audio_size_bytes": ep.audio_size_bytes,
            "created_at": ep.created_at,
            "steps": {},
            "total_input_tokens": 0,
            "total_output_tokens": 0,
            "total_cost": 0.0,
            "total_duration_seconds": 0.0,
        }

        for job in ep.jobs:
            if job.status != "completed":
                if job.started_at and job.completed_at:
                    wall = (job.completed_at - job.started_at).total_seconds()
                    row["steps"][job.step] = {"wall_seconds": round(wall, 2)}
                    row["total_duration_seconds"] += wall
                continue

            metrics = json.loads(job.metrics_json) if job.metrics_json else {}
            step_data = dict(metrics)

            if job.started_at and job.completed_at:
                wall = (job.completed_at - job.started_at).total_seconds()
                step_data["wall_seconds"] = round(wall, 2)
                row["total_duration_seconds"] += wall

            row["steps"][job.step] = step_data

            input_t = metrics.get("input_tokens", 0)
            output_t = metrics.get("output_tokens", 0)
            row["total_input_tokens"] += input_t
            row["total_output_tokens"] += output_t
            totals["total_input_tokens"] += input_t
            totals["total_output_tokens"] += output_t

            if job.step == "tts":
                totals["total_tts_seconds"] += metrics.get("duration_seconds", 0)

        # Accumulate per-job cost using per-model pricing
        for job in ep.jobs:
            if job.status != "completed" or not job.metrics_json:
                continue
            m = json.loads(job.metrics_json) if job.metrics_json else {}
            in_t = m.get("input_tokens", 0)
            out_t = m.get("output_tokens", 0)
            model_name = m.get("model", "")
            row["total_cost"] += _calc_cost(in_t, out_t, model_name)
        totals["total_cost"] += row["total_cost"]
        totals["total_generation_seconds"] += row["total_duration_seconds"]

        if ep.audio_duration_seconds:
            totals["total_audio_seconds"] += ep.audio_duration_seconds

        episode_rows.append(row)

    return templates.TemplateResponse(
        request, "metrics.html", context={"totals": totals, "episodes": episode_rows}
    )

File: src/routes/pages.ts
import type { FastifyInstance } from "fastify";
import { eq, desc } from "drizzle-orm";
import { db } from "../database.js";
import { episodes, podcastSettings, type User } from "../schema.js";
import { requireAuthPage, requireAdminPage } from "../auth.js";
import {
  createEpisode,
  getEpisode,
  listEpisodes,
} from "../services/episode.js";
import { getTtsProgress as readTtsProgress } from "../services/tts.js";
import { getAllModelPricing } from "../services/llm-providers.js";
import { DEFAULT_TONE_NOTES } from "../services/transcript.js";
import { renderTemplate } from "../server.js";

const MODEL_PRICING = getAllModelPricing();
const DEFAULT_PRICING = { input: 3.0, output: 15.0 };

// ─── Template helpers ────────────────────────────────────────────────
export function statusBadge(status: string): string {
  const mapping: Record<string, string> = {
    ready: "badge--success",
    failed: "badge--error",
    pending: "badge--info",
  };
  return mapping[status] ?? "badge--warning";
}

export function statusLabel(status: string): string {
  const mapping: Record<string, string> = {
    pending: "Pending",
    researching: "Researching",
    writing_transcript: "Writing transcript",
    generating_audio: "Generating audio",
    encoding: "Encoding",
    ready: "Ready",
    failed: "Failed",
  };
  return mapping[status] ?? status.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}

export function formatDuration(seconds: number | null): string {
  if (seconds == null) return "";
  const minutes = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return `${minutes}:${String(secs).padStart(2, "0")}`;
}

export function formatFileSize(sizeBytes: number | null): string {
  if (sizeBytes == null) return "";
  if (sizeBytes < 1024) return `${sizeBytes} B`;
  if (sizeBytes < 1024 * 1024)
    return `${(sizeBytes / 1024).toFixed(1)} KB`;
  return `${(sizeBytes / (1024 * 1024)).toFixed(1)} MB`;
}

const PIPELINE_STEPS = ["research", "transcript", "tts", "encode"];
const STEP_LABELS: Record<string, string> = {
  research: "Research",
  transcript: "Transcript",
  tts: "Audio",
  encode: "Encode",
};

export function buildPipelineInfo(episode: any): Array<Record<string, unknown>> {
  const jobsByStep: Record<string, any> = {};
  for (const job of episode.jobs ?? []) {
    jobsByStep[job.step] = job;
  }

  return PIPELINE_STEPS.map((stepName) => {
    const job = jobsByStep[stepName];
    if (job) {
      let duration = null;
      if (job.startedAt && job.completedAt) {
        const totalSecs = Math.round(
          (new Date(job.completedAt).getTime() -
            new Date(job.startedAt).getTime()) /
            1000
        );
        duration =
          totalSecs >= 60
            ? `${Math.floor(totalSecs / 60)}m ${totalSecs % 60}s`
            : `${totalSecs}s`;
      }
      return {
        name: stepName,
        label: STEP_LABELS[stepName],
        status: job.status,
        attempts: job.attempts,
        duration,
      };
    }
    return {
      name: stepName,
      label: STEP_LABELS[stepName],
      status: "waiting",
      attempts: 0,
      duration: null,
    };
  });
}

export function getCurrentStepIndex(episode: any): number {
  if (episode.status === "ready") return PIPELINE_STEPS.length;
  const jobsByStep: Record<string, any> = {};
  for (const job of episode.jobs ?? []) {
    jobsByStep[job.step] = job;
  }
  for (let i = 0; i < PIPELINE_STEPS.length; i++) {
    const job = jobsByStep[PIPELINE_STEPS[i]];
    if (!job || job.status === "pending" || job.status === "running") return i;
  }
  return PIPELINE_STEPS.length;
}

export function getTtsProgress(episode: any): Record<string, unknown> | null {
  if (episode.status !== "generating_audio") return null;
  const progress = readTtsProgress(episode.id);
  if (!progress) return null;
  return progress;
}

function calcCost(
  inputTokens: number,
  outputTokens: number,
  model = ""
): number {
  const pricing = MODEL_PRICING[model] ?? DEFAULT_PRICING;
  return (
    (inputTokens * pricing.input + outputTokens * pricing.output) / 1_000_000
  );
}

function html(reply: any, content: string) {
  return reply.header("Content-Type", "text/html").send(content);
}

export async function pageRoutes(app: FastifyInstance) {
  // Shoo callback
  app.get("/shoo/callback", async (_request, reply) => {
    return html(reply, renderTemplate("auth_callback.html", {}));
  });

  // Home — new episode form
  app.get("/", async (request, reply) => {
    const user = await requireAuthPage(request, reply);
    return html(reply, renderTemplate("episode_new.html", { user }));
  });

  // Create episode via form POST
  app.post("/", async (request, reply) => {
    const user = await requireAuthPage(request, reply);
    const form = request.body as Record<string, string>;
    const topic = (form.topic || "").trim();
    const title = (form.title || "").trim() || null;
    let targetLength = parseInt(form.target_length_minutes || "30", 10);
    if (![15, 30, 60, 120].includes(targetLength)) targetLength = 30;

    if (!topic) {
      return html(
        reply,
        renderTemplate("episode_new.html", {
          error: "Topic is required",
          user,
        })
      );
    }

    const episode = await createEpisode(
      topic,
      title,
      null,
      targetLength,
      user.id
    );
    return reply.code(303).redirect( `/episodes/${episode.id}`);
  });

  // Also handle POST /episodes/new for the form
  app.post("/episodes/new", async (request, reply) => {
    const user = await requireAuthPage(request, reply);
    const form = request.body as Record<string, string>;
    const topic = (form.topic || "").trim();
    const title = (form.title || "").trim() || null;
    let targetLength = parseInt(form.target_length_minutes || "30", 10);
    if (![15, 30, 60, 120].includes(targetLength)) targetLength = 30;

    if (!topic) {
      return html(
        reply,
        renderTemplate("episode_new.html", {
          error: "Topic is required",
          user,
        })
      );
    }

    const episode = await createEpisode(
      topic,
      title,
      null,
      targetLength,
      user.id
    );
    return reply.code(303).redirect( `/episodes/${episode.id}`);
  });

  // Episode list
  app.get("/episodes", async (request, reply) => {
    const user = await requireAuthPage(request, reply);
    const eps = await listEpisodes(user.id);
    return html(
      reply,
      renderTemplate("index.html", { episodes: eps, user })
    );
  });

  // Legacy redirect
  app.get("/episodes/new", async (_request, reply) => {
    return reply.code(301).redirect( "/");
  });

  // Episode detail
  app.get<{ Params: { episodeId: string } }>(
    "/episodes/:episodeId",
    async (request, reply) => {
      const user = await requireAuthPage(request, reply);
      const episode = await getEpisode(request.params.episodeId, user.id);
      if (!episode) {
        return reply.code(404).header("Content-Type", "text/html").send("Not found");
      }
      return html(
        reply,
        renderTemplate("episode_detail.html", { episode, user })
      );
    }
  );

  // Logs page (admin)
  app.get("/logs", async (request, reply) => {
    const user = await requireAdminPage(request, reply);
    return html(reply, renderTemplate("logs.html", { user }));
  });

  // Settings page
  app.get("/settings", async (request, reply) => {
    const user = await requireAuthPage(request, reply);
    let [s] = await db
      .select()
      .from(podcastSettings)
      .where(eq(podcastSettings.userId, user.id))
      .limit(1);
    if (!s) {
      [s] = await db
        .insert(podcastSettings)
        .values({ userId: user.id })
        .returning();
    }

    let toneNotes = [...DEFAULT_TONE_NOTES];
    if (s.transcriptToneNotes) {
      try {
        const parsed = JSON.parse(s.transcriptToneNotes);
        if (Array.isArray(parsed)) toneNotes = parsed;
      } catch {}
    }

    return html(
      reply,
      renderTemplate("settings.html", { settings: s, tone_notes: toneNotes, user })
    );
  });

  // Settings form POST
  app.post("/settings", async (request, reply) => {
    const user = await requireAuthPage(request, reply);
    const form = request.body as Record<string, string>;

    let [s] = await db
      .select()
      .from(podcastSettings)
      .where(eq(podcastSettings.userId, user.id))
      .limit(1);
    if (!s) {
      [s] = await db
        .insert(podcastSettings)
        .values({ userId: user.id })
        .returning();
    }

    const updateData: Record<string, unknown> = {};
    const fieldMap: Record<string, string> = {
      title: "title",
      description: "description",
      author: "author",
      language: "language",
      host_a_name: "hostAName",
      host_b_name: "hostBName",
    };

    for (const [formField, dbField] of Object.entries(fieldMap)) {
      const value = (form[formField] || "").trim();
      if (value) updateData[dbField] = value;
    }

    if (Object.keys(updateData).length > 0) {
      await db
        .update(podcastSettings)
        .set(updateData)
        .where(eq(podcastSettings.userId, user.id));
    }

    return reply.code(303).redirect( "/settings");
  });

  // Metrics page (admin)
  app.get("/metrics", async (request, reply) => {
    const user = await requireAdminPage(request, reply);

    const allEpisodes = await db.query.episodes.findMany({
      with: { jobs: true },
      orderBy: [desc(episodes.createdAt)],
    });

    const totals = {
      episodes: 0,
      episodes_ready: 0,
      total_input_tokens: 0,
      total_output_tokens: 0,
      total_cost: 0,
      total_audio_seconds: 0,
      total_generation_seconds: 0,
      total_tts_seconds: 0,
    };

    const episodeRows = allEpisodes.map((ep) => {
      totals.episodes++;
      if (ep.status === "ready") totals.episodes_ready++;

      const row: Record<string, unknown> = {
        id: ep.id,
        title: ep.title,
        status: ep.status,
        episode_number: ep.episodeNumber,
        audio_duration_seconds: ep.audioDurationSeconds,
        audio_size_bytes: ep.audioSizeBytes,
        created_at: ep.createdAt,
        steps: {} as Record<string, unknown>,
        total_input_tokens: 0,
        total_output_tokens: 0,
        total_cost: 0,
        total_duration_seconds: 0,
      };

      for (const job of ep.jobs) {
        if (job.status !== "completed" || !job.metricsJson) {
          if (job.startedAt && job.completedAt) {
            const wall =
              (job.completedAt.getTime() - job.startedAt.getTime()) / 1000;
            (row.steps as any)[job.step] = {
              wall_seconds: Math.round(wall * 100) / 100,
            };
            (row.total_duration_seconds as number) += wall;
          }
          continue;
        }

        const metrics = JSON.parse(job.metricsJson);
        const stepData = { ...metrics };

        if (job.startedAt && job.completedAt) {
          const wall =
            (job.completedAt.getTime() - job.startedAt.getTime()) / 1000;
          stepData.wall_seconds = Math.round(wall * 100) / 100;
          (row.total_duration_seconds as number) += wall;
        }

        (row.steps as any)[job.step] = stepData;

        const inputT = metrics.input_tokens ?? 0;
        const outputT = metrics.output_tokens ?? 0;
        (row.total_input_tokens as number) += inputT;
        (row.total_output_tokens as number) += outputT;
        totals.total_input_tokens += inputT;
        totals.total_output_tokens += outputT;

        if (job.step === "tts") {
          totals.total_tts_seconds += metrics.duration_seconds ?? 0;
        }
      }

      // Cost
      for (const job of ep.jobs) {
        if (job.status !== "completed" || !job.metricsJson) continue;
        const m = JSON.parse(job.metricsJson);
        (row.total_cost as number) += calcCost(
          m.input_tokens ?? 0,
          m.output_tokens ?? 0,
          m.model ?? ""
        );
      }
      totals.total_cost += row.total_cost as number;
      totals.total_generation_seconds += row.total_duration_seconds as number;
      if (ep.audioDurationSeconds)
        totals.total_audio_seconds += ep.audioDurationSeconds;

      return row;
    });

    return html(
      reply,
      renderTemplate("metrics.html", {
        totals,
        episodes: episodeRows,
        user,
      })
    );
  });
}
